File: ios/PulseWatchKit/Sources/PulseWatchAPI/EnrollmentFlow.swift
import Foundation
import CryptoKit
import PulseWatchModels
import PulseWatchVault

/// High-level helpers that compose the API client with the vault — the things
/// the UI actually wants.
public final class EnrollmentFlow: @unchecked Sendable {
    private let client: APIClient
    public init(client: APIClient) { self.client = client }

    /// Performs `POST /v1/auth/devices` and persists the bearer token. Returns
    /// the device id so the caller can show it on the device-management screen.
    public func enrollDevice(
        platform: DevicePlatform,
        deviceName: String?,
        pairingCode: String? = nil
    ) async throws -> String {
        let kp = Curve25519.KeyAgreement.PrivateKey()
        let s = try await client.enrollDevice(
            platform: platform,
            publicKey: kp.publicKey.rawRepresentation,
            pairingCode: pairingCode,
            deviceName: deviceName
        )
        let expires = ISO8601DateFormatter().date(from: s.expires_at) ?? Date(timeIntervalSinceNow: 90 * 86400)
        try client.session.save(.init(deviceId: s.device_id, token: s.token, expiresAt: expires, baseURL: client.baseURL))
        return s.device_id
    }

    public func enrollPhone(deviceName: String?) async throws -> String {
        try await enrollDevice(platform: .ios, deviceName: deviceName)
    }

    /// Wraps the provider key for the backend and creates the account. The
    /// plaintext provider key is zeroed before this method returns.
    public func enrollProviderAccount(
        provider: Provider,
        label: String,
        providerKey: String,
        orgId: String? = nil
    ) async throws -> Account {
        let wrap = try await client.currentWrappingKey()
        guard let recipient = Data(base64Encoded: wrap.public_key) else {
            throw APIError.http(status: -1, body: "bad wrapping key")
        }
        var keyBytes = Data(providerKey.utf8)
        defer { keyBytes.resetBytes(in: 0..<keyBytes.count) }

        let envelope = try EnvelopeCipher.seal(plaintext: keyBytes, recipientRawPublicKey: recipient)
        return try await client.createAccount(.init(
            provider: provider,
            label: label,
            wrapped_key: envelope.base64EncodedString(),
            kid: wrap.kid,
            org_id: orgId
        ))
    }
}

File: android/core/src/main/java/app/cap/core/api/EnrollmentFlow.kt
package app.cap.core.api

import android.util.Base64
import app.cap.core.vault.EnvelopeCipher
import app.cap.core.vault.SessionStore
import app.cap.core.vault.StoredSession
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import java.security.SecureRandom

class EnrollmentFlow(private val client: ApiClient) {

    suspend fun enrollDevice(
        platform: DevicePlatform,
        deviceName: String?,
        pairingCode: String? = null,
    ): String {
        // X25519 keypair carried over the enrollment for forward use; M3-equivalent
        // of the iOS phone↔watch handoff.
        val priv = X25519PrivateKeyParameters(SecureRandom())
        val pub = ByteArray(32).also { priv.generatePublicKey().encode(it, 0) }

        val session = client.enrollDevice(
            DeviceEnrollRequest(
                platform = platform,
                public_key = Base64.encodeToString(pub, Base64.NO_WRAP),
                pairing_code = pairingCode,
                device_name = deviceName,
            )
        )
        client.sessionStore.save(
            StoredSession(
                deviceId = session.device_id,
                token = session.token,
                expiresAt = session.expires_at,
                baseUrl = client.baseUrl,
            )
        )
        return session.device_id
    }

    /**
     * Wraps the provider key for the backend's current X25519 wrapping public
     * key and creates the account. The plaintext byte array is zeroed before
     * this method returns.
     */
    suspend fun enrollProviderAccount(
        provider: Provider,
        label: String,
        providerKey: String,
        orgId: String? = null,
    ): Account {
        val wrap = client.currentWrappingKey()
        val recipient = Base64.decode(wrap.public_key, Base64.NO_WRAP)
        val plaintext = providerKey.toByteArray(Charsets.UTF_8)
        try {
            val envelope = EnvelopeCipher.seal(plaintext, recipient)
            return client.createAccount(
                AccountCreate(
                    provider = provider,
                    label = label,
                    wrapped_key = Base64.encodeToString(envelope, Base64.NO_WRAP),
                    kid = wrap.kid,
                    org_id = orgId,
                )
            )
        } finally {
            java.util.Arrays.fill(plaintext, 0)
        }
    }
}
